### ISSUE TITLE
Denied PUBLISH after killall -HUP

### ISSUE THREAD

**Issue by brocaar** [eclipse/mosquitto#1277]
Recently we noticed that all published messages from our MQTT clients were rejected with a _Denied PUBLISH from_ entry in the logfile. Restarting the Mosquitto process helped, for a while. We are not sure what triggered this, potentially this started to happen after installing security patches. Before this instance was working fine. We did not make any configuration changes. 

After some debugging, we noticed that this issue occurred every time the logrotate was triggered. The Mosquitto logrotate config contains:

```
        postrotate
                /usr/bin/killall -HUP mosquitto
        endscript
```

Executing `/usr/bin/killall -HUP mosquitto` manually triggered the same behavior. Before executing this command, all clients are successfully publishing messages. After execution, publish commands are rejected with _Denied PUBLISH from_ ...

```
1557469417: Received PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
1557469447: Received PINGREQ from 804985b3-5722-4eba-803e-1452d7ea8a65
1557469447: Sending PINGRESP to 804985b3-5722-4eba-803e-1452d7ea8a65
1557469447: Received PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
1557469477: Received PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))

-- logrotate happens here --

1557469507: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
1557469537: Received PINGREQ from 804985b3-5722-4eba-803e-1452d7ea8a65
1557469537: Sending PINGRESP to 804985b3-5722-4eba-803e-1452d7ea8a65
1557469537: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
1557469559: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (426 bytes))
1557469563: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (426 bytes))
1557469567: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
1557469567: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (426 bytes))
1557469597: Denied PUBLISH from 804985b3-5722-4eba-803e-1452d7ea8a65 (d0, q0, r0, m0, 'topic', ... (221 bytes))
```

---

### Environment

Mosquitto version: tested with 1.4.8 (Ubuntu repo) & 1.6.2 (from PPA)
Linux: Ubuntu 16.04.6 LTS

Configuration files:

#### `/etc/mosquitto/mosquitto.conf`

```
pid_file /var/run/mosquitto.pid

persistence false
persistence_location /var/lib/mosquitto/

log_dest file /var/log/mosquitto/mosquitto.log

include_dir /etc/mosquitto/conf.d
```

#### `/etc/mosquitto/conf.d/local.conf`

```
connection_messages true

log_type all

allow_anonymous false
password_file /etc/mosquitto/pwd
acl_file /etc/mosquitto/acl

bind_address a.b.c.d
listener 5088

cafile /path/to/ca...
certfile /path/to/cert...
keyfile /path/to/key...
```

---

### Reproducing

I've been trying to reproduce this issue on a different server (clean VM) and the strange thing is, I can not reproduce this. Maybe this is related to the number of MQTT clients connected? Maybe this is a race-condition somewhere in the code that triggers this behavior of Mosquitto? I'm not sure... However, I can consistently reproduce this on one or our instances.

Searching in the issues on GitHub, I found #986, #896 and #657 which potentially could be related.

**Comment by brocaar** [comment]
While writing and providing more context, I found the actual issue which caused this. This might be helpful for others:

The `/etc/mosquitto/pwd` file was owned by `root:root` and only `root` had read/write access to it. For some reason, Mosquitto would start fine and was able to read this. However, `/usr/bin/killall -HUP mosquitto` would throw a _Error: Unable to open pwfile_ error in the logs.

After chowning the `pwd` file to `mosquitto:mosquitto` all worked fine. I'm not sure why this issue started after an `apt update && apt upgrade`. Maybe something changed related to how Mosquitto is started (Systemd or `/etc/init.d/mosquitto` scripts) or maybe the update did update Mosquitto and there was a change in the source related to this? 

Anway, I found it quite confusing that a (re)start of Mosquitto does read the password file, but a `/usr/bin/killall -HUP mosquitto` fails. Maybe this is something that could be improved?
